bitcoin/src/address.rs

```diff
@@ -727,34 +820,32 @@ impl Address {
         }
     }
 
-    /// Returns true if the given pubkey is directly related to the address payload.
+    /// Checks whether network of this address is as expected.
     ///
-    /// This is determined by directly comparing the address payload with either the
-    /// hash of the given public key or the segwit redeem hash generated from the
-    /// given key. For taproot addresses, the supplied key is assumed to be tweaked
-    pub fn is_related_to_pubkey(&self, pubkey: &PublicKey) -> bool {
-        let pubkey_hash = pubkey.pubkey_hash();
-        let payload = self.payload.as_bytes();
-        let xonly_pubkey = XOnlyPublicKey::from(pubkey.inner);
-
-        (*pubkey_hash == *payload)
-            || (xonly_pubkey.serialize() == *payload)
-            || (*segwit_redeem_hash(&pubkey_hash) == *payload)
+    /// For details about this mechanism, see section *Parsing addresses* on [`Address'].
+    #[inline]
+    pub fn expect_network(self, expected: Network) -> Result<Address, Error> {
+        if self.is_valid_for_network(expected) {
+            Ok(self.unsafe_mark_network_valid())
+        } else {
+            Err(Error::NetworkValidation { found: self.network, expected })
+        }
     }
 
-    /// Returns true if the supplied xonly public key can be used to derive the address.
+    /// Marks, without any checks, network of this address as valid.
     ///
-    /// This will only work for Taproot addresses. The Public Key is
-    /// assumed to have already been tweaked.
-    pub fn is_related_to_xonly_pubkey(&self, xonly_pubkey: &XOnlyPublicKey) -> bool {
-        let payload = self.payload.as_bytes();
-        payload == xonly_pubkey.serialize()
+    /// This is a life-threatening method, reader will most likely prefer [`expected_network`]
```

Comment:
Seems needlessly scary. I would simply say "Invalid use may lead to loss of funds!"

Reply:
Right, got carried away a bit. I rephrased it in 0e448ae3398cd4486855693f524c7e0e40f7f902.